Make <caml/sync.h> more abstract and refactor implementation of sync.c

xavierleroy opened the pull request:
## Current situation
- <caml/sync.h> defines some operations over the POSIX implementation of mutexes and condition variables
- These operations are used in sync.c to provide the full POSIX implementation of mutexes and condition variables
- They are also used in domain.c to lock and unlock the termination mutex associated with a domain
- They are also used in otherlibs/systhreads/st_posix.h for the POSIX implementation of the threads library.

## The problem
- <caml/sync.h> is installed along with the other <caml/*> headers, yet it's not protected by `CAML_INTERNALS` and doesn't follow the `caml_` naming convention for exported things.
- <caml/sync.h> and its users are (too?) specific to POSIX threads, which will cause issues if we ever decide to provide a pure Win32 implementation of mutexes and condition variables.

## The proposed solution
- <caml/sync.h> declares only the high-level, platform-agnostic mutex and condition variable operations that operate on OCaml values.  The declarations are currently made `CAML_INTERNALS` but we could envision other C/OCaml interface code to use them.
- domain.c is adjusted to use these high-level operations when acting on domain termination mutexes.  It is as GC-safe as the previous implementation, since `ml_values->mutex` is registered as a root.
- The POSIX implementation of mutexes and condition variables that used to be in <caml/sync.h> is moved to sync_posix.h (an unexported header, included in sync.c only) and extended with all POSIX-specific operations.
- A future Win32 implementation could, then, go in sync_win32.h.
- sync.c and otherlibs/systhreads/st_posix.h are adjusted accordingly.

This was discussed briefly with @ctk21 beforehand.

xavierleroy committed:
Make <caml/sync.h> more abstract and refactor implementation of sync.c

<caml/sync.h> declares only the high-level, platform-agnostic mutex
and condition variable operations that operate on OCaml values.

domain.c is adjusted to use these high-level operations to work with
termination mutexes.

The POSIX implementation of mutexes and condition variables that used
to be in <caml/sync.h> is moved to sync_posix.h (an unexported header,
included in sync.c only) and extended with all POSIX-specific
operations.

sync.c and otherlibs/systhreads/st_posix.h are adjusted accordingly.

avsm commented on runtime/sync_posix.h:
```diff
@@ -0,0 +1,140 @@
+  if (rc != 0) goto error1;
+  rc = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+  if (rc != 0) goto error2;
+  m = caml_stat_alloc_noexc(sizeof(pthread_mutex_t));
```
Probably orthogonal to this PR, but can't the memory allocation be moved up to the top and just use `caml_stat_alloc` to reduce the various `goto error` paths later?  caml_stat_alloc will raise Out_of_memory immediately, just as sync_error does later on anyway.

xavierleroy replied:
This code is taken verbatim from otherlibs/systhreads/st_posix.h in OCaml 4, so I'd rather not change it in this PR, just trying to find a good place for this code.

avsm commented on runtime/sync_posix.h:
```diff
@@ -0,0 +1,140 @@
+{
+  int rc;
+  sync_condvar c = caml_stat_alloc_noexc(sizeof(pthread_cond_t));
+  if (c == NULL) return ENOMEM;
```
Similar question about caml_stat_alloc here and directly raising Out_of_memory.

ctk21 commented on runtime/domain.c:
```diff
@@ -845,7 +843,7 @@ static void* domain_thread_func(void* v)
     /* Joining domains will lock/unlock the terminate_mutex so this unlock will
        release them if any domains are waiting. */
-    sync_mutex_unlock(terminate_mutex);
+    caml_ml_mutex_unlock(ml_values->mutex);
```
It is worth noting a subtlety here with the garbage collection.

At this point, the domain has left the STW set. It can not read any OCaml values which might be moved or collected by the garbage collector. So why is it safe to read `ml_values->mutex`?

I believe this argument shows safety:
- (as things stand) values are only moved by the minor collector and `domain_terminate` forces all minor heaps to have been collected once. Therefore `ml_values->mutex` must be in the major heap and will not be moved.
- we have registered `ml_values->mutex` as a global root and so therefore it can not be collected from the major heap while we unlock it.
- we are operating on the global roots _from a thread outside the STW set_. This relies on the fact that the global root updates and global root scanning are protected from each other by locks in the global roots. That is the free of `ml_values` can not happen while the roots are being scanned from within a STW. 

An alternative design would be to unlock during `domain_terminate` when we are about to drop out of the STW set. Notice that after leaving the STW set the terminating domain then goes on to do cleanup (e.g. deallocate memory, shutdown the backup thread, etc.). The more subtle design allows the unlock of the terminate mutex to include all the domain cleanup processing. 

```suggestion
    /* The domain has left the STW set, but all minor heaps will have been
       collected in domain_terminate.
       So ml_values->mutex is in the major heap and will not be moved */
    CAMLassert(!Is_young(ml_values->mutex));
    caml_ml_mutex_unlock(ml_values->mutex);
```

xavierleroy replied:
Thanks a lot for the detailed analysis and explanations!  My mental picture was less detailed, as I viewed the Custom heap block and the sync_mutex pointer it contains as "equally valid" (i.e. both are valid or both were deallocated by the GC).  But that's not true if a minor GC is concurrently moving heap blocks, indeed.

I'll merge your comment in this PR.  I'll look into unlocking in `domain_terminate`.

damiendoligez replied:
I'm a bit worried that this will become false if we someday reimplement heap compaction.
I don't think we should change this code to prepare for that, but maybe add a comment so that whoever implements compaction has a chance to learn about the problem?

xavierleroy replied:
Tentative implementation of unlocking in `domain_terminate`  in 2db69380a .  I tried to release the termination lock "as late as possible".  I'm not quite sure I know what I'm doing here :-)

ctk21 replied:
I think https://github.com/ocaml/ocaml/commit/2db69380aa9193c753eddd2fc60ef11a8137453d is sound. While holding the `all_domains_lock`, no new STW section can start (as in the comment around L1361) and so no minor or major garbage collections can initiate while this domain drops itself from the STW set. 

The niggle is the domain join completes but the terminating domain continues with runtime cleanup; however the terminating domain never runs OCaml code after that point. We could decide this is fine. 

There is a third option: pull the custom sync_mutex pointer out of the OCaml block in a stack variable inside `domain_thread_func` (i.e. the variable `terminate_mutex` in the original code). This sync_mutex pointer can not be moved by the garbage collector and so liveness is the only concern which is handled by the global root. 

Advantages: 
- we unblock joining threads at the very end (maybe OCaml based timing of domain terminate would be more accurate? maybe tighter resource usage?)

Disadvantages: 
- needing to provide a sync interface both for the ML level and operating on the pointers in the cusom blocks 
- the more complex soundness argument for the code might be a liability

I'm leaning in favour of the simplicity of https://github.com/ocaml/ocaml/commit/2db69380aa9193c753eddd2fc60ef11a8137453d  until there is a use case problem with the cleanup code and join.

xavierleroy replied:
> until there is a use case problem with the cleanup code and join.

I can think of a possible problem: in "cleanup at exit" mode, if other domains or threads are still running, memory can be freed "under their feet" and cause crashes.  So the advice is to join all threads or domains before the end of the initial domain.  But here some domains may still be running cleanup code after the join, which could interfere with the "cleanup at exit".

I'll give your third option a try.  The <caml/sync.h> interface will need a system-dependent definition of `sync_mutex`, but that should be doable.

ctk21 reviewed (approved):
I think the motivation is sensible and fixes things. I have taken a look and believe it to be implementing what it intends.  

I needed to spend some time convincing myself the termination unlock operation was safe with respect to the garbage collector. A comment and/or assert might help there.

xavierleroy committed:
Explain why `ml_values->mutex` is safe to access when domain terminates

Co-authored-by: Tom Kelly <ctk21@users.noreply.github.com>

xavierleroy committed:
Experiment: release the termination lock in domain_terminate

So that we access the heap-allocated lock while still among the STW
participants.  This feels safer.

gasche commented:
I wonder if @ctk21 or @damiendoligez would want to approve of this PR.

xavierleroy commented:
Wait, I'm still working on it.

xavierleroy committed:
Minimize changes to domain.c

This commit reverts most of the changes to domain.c, so that
the termination lock is released late and manipulated as a
pointer outside the heap, like before.  It looks safer this way.

To support this, <caml/sync.h> is changed once more to export
the type `sync_mutex` and two lock / unlock functions defined in sync.c
and usable in domain.c.

xavierleroy commented:
I just pushed approach number 3, which reverts essentially all changes to domain.c and exposes again the type `sync_mutex` in `<caml/sync.h>`.

xavierleroy commented:
Could the latest version of this PR get a review, please? It's smaller than the first version, doesn't change the logic in domain.c, just moves definitions around header files to improve name hygiene and future Win32 portability.  Thanks.

ctk21 commented on runtime/domain.c:
```diff
@@ -851,7 +851,7 @@ static void* domain_thread_func(void* v)
     terminate_mutex = Mutex_val(ml_values->mutex);
     /* we lock terminate_mutex here and unlock when the domain is torn down
       this provides a simple block for domains attempting to join */
```
```suggestion
      this provides a simple block for domains attempting to join 
      NB: terminate_mutex will not be moved by the garbage collector
      as it is not an OCaml block. ml_values->mutex is registered as
      a global root and keeps the mutex custom memory alive with 
      the garbage collector. 
```

xavierleroy replied:
More comments is good!  Thanks.

ctk21 reviewed (approved):
I believe the PR to be sensible and that it does what it intends.

I've made a suggestion to add a comment for `terminate_mutex`; 
my aim was to capture the end result of the PR discussion in the
code.
Feel free to ignore the suggestion if you feel it doesn't add.

gasche reviewed (approved):
Approving on @ctk21's behalf.

xavierleroy commented:
Thank you! Time for squash and merge...

xavierleroy committed:
Comment on GC safety of terminate_mutex.

Co-authored-by: Tom Kelly <ctk21@users.noreply.github.com>

shindere commented:
Just pushed e228907f21b7e172e6b3c1a259f5ba1a8697be48 to please check-typo
